add support for any config parameter

cb-jake opened the pull request:
### _Summary_

This PR extends the preferences object to support any config value. This PR also exposes new methods to get the provider or an new sdk object. 

### _How did you test your changes?_

Manually. In a follow up PR there will be an update to the test app to use the new create SDK object

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

fan-zhang-sv commented on packages/wallet-sdk/src/core/provider/interface.ts:
```diff
@@ -42,10 +42,47 @@ export interface AppMetadata {
+  /**
+   * @deprecated this is only for internal use
+   */
   keysUrl?: string;
```
since this is only used by us, can we remove it for now

if anything, we will always catch the problem before we roll out as this is used internally i think?

cb-jake replied:
Might be good to list it as deprecated for a few versions just in case someone externally is using it

fan-zhang-sv replied:
hmm prob there won't be any one using it externally, let's remove it?

cb-jake replied:
Sounds good

fan-zhang-sv reviewed (commented):
looking good, left a comment!

spencerstock commented on packages/wallet-sdk/src/core/provider/interface.ts:
```diff
@@ -42,10 +42,47 @@ export interface AppMetadata {
+   * @type {PostOnboardingAction}
+   * @description Recommends the action to be taken after the onboarding process to engage users with various flows. Possible values:
+   * - `none`: No action is recommended post-onboarding. (Default experience)
+   * - `onramp`: Recommends initiating the onramp flow, allowing users to prefill their account with a specified asset, chain, and amount.
```
> allowing users to prefill their account with a specified asset, chain, and amount.

"allowing users to prefill their account with an optional <OnrampPrefillOptions>"

spencerstock commented on packages/wallet-sdk/src/core/provider/interface.ts:
```diff
@@ -42,10 +42,47 @@ export interface AppMetadata {
+  /**
+   * @param postOnboardingAction
+   * @type {PostOnboardingAction}
+   * @description Recommends the action to be taken after the onboarding process to engage users with various flows. Possible values:
```
> Recommends the action to be taken after the onboarding process to engage users with various flows. Possible values:

I think we could add some clarity here
"Displays CTAs to the user based on the preference of the dapp. These CTAs are part of prebuilt UI components that are available to the CB Smart Wallet"

cb-jake replied:
Thanks good callout

spencerstock commented on packages/wallet-sdk/src/core/provider/interface.ts:
```diff
@@ -42,10 +42,47 @@ export interface AppMetadata {
+   * Expects a 4 byte hex string
+   */
+  attributionDataSuffix?: string;
+} & Record<string, unknown>;
```
nice

spencerstock commented on packages/wallet-sdk/src/core/provider/interface.ts:
```diff
@@ -42,10 +42,47 @@ export interface AppMetadata {
+   */
+  postOnboardingAction?: PostOnboardingAction;
+  /**
+   * @param onrampPrefillOptions
+   * @type {OnrampPrefillOptions}
+   * @description This option only functions when `postOnboardingAction` is set to `onramp`.
+   * - Prefills the onramp flow with the specified asset, chain, and suggested amount,
+   * allowing users to prefill their account.
+   * - Ensure the asset and chain are supported by the onramp provider
+   * (e.g., Coinbase Pay - CBPay).
+   */
```
add: defaults to an asset selector with 0 as the initial amount

Can we also add some help for devs in determining the right formats here, and determining what is supported by CBPay?

cb-jake replied:
updated with a link from cbpay docs

spencerstock reviewed (commented):
Looks good! 
1. How do we prevent dapps from mistakenly passing these in with eoa connections and then thinking something is broken when they do not make a difference?
2. I see some of the language is staying generic, i.e. other implementers could handle these functions differently. I think we might be better off expicitly coupling these to SCW and having the descriptions be clearer at the cost of future extensibility

cb-jake commented:
@spencerstock 
> How do we prevent dapps from mistakenly passing these in with eoa connections and then thinking something is broken when they do not make a difference?

I think documentation makes the most sense. Inside the doc blocks I've added a Smart Wallet Only tag

>I see some of the language is staying generic, i.e. other implementers could handle these functions differently. I think we might be better off expicitly coupling these to SCW and having the descriptions be clearer at the cost of future extensibility

Agreed. I've updated the jsdocs

cb-jake commented on packages/wallet-sdk/src/CoinbaseWalletProvider.ts:
```diff
@@ -27,10 +28,14 @@ export class CoinbaseWalletProvider extends ProviderEventEmitter implements Prov
     this.preference = preference;
-    this.communicator = new Communicator(keysUrl, metadata);
+
+    const url = typeof keysUrl === 'string' ? keysUrl : CB_KEYS_URL;
```
Removing keysUrl as an option resulted in a type error here. Now we're type checking for a string before passing to the communicator. Might be worth allowing unknown and validating inside the communicator

fan-zhang-sv replied:
i see, actually maybe good to have a separate PR to remove keysUrl? i can take it on, P2/P3 maybe?

fan-zhang-sv replied:
we can revert all keysUrl change in this PR? i'll tackle it separately

fan-zhang-sv reviewed (dismissed):
ok to merge also, i can take remove the usage of keysUrl in a follow up PR separately

fan-zhang-sv reviewed (approved):
ty! lfg
